cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageTest.java

```diff
@@ -77,13 +87,120 @@ public void testLineage() throws Exception {
+                                         ImmutableSet.of(Id.Flow.Flowlet.from(flow, AllProgramsApp.A.NAME))),
+                            new Relation(dataset, mapreduce, AccessType.UNKNOWN, ImmutableSet.of(mrRunId)),
+                            new Relation(dataset, spark, AccessType.UNKNOWN, ImmutableSet.of(sparkRunId)),
+                            new Relation(dataset, mapreduce, AccessType.UNKNOWN, ImmutableSet.of(workflowMrRunId)),
+                            new Relation(dataset, service, AccessType.UNKNOWN, ImmutableSet.of(serviceRunId)),
+                            new Relation(dataset, worker, AccessType.UNKNOWN, ImmutableSet.of(workerRunId))
+                          ));
+      Assert.assertEquals(expected, lineage);
+    } finally {
+      try {
+        deleteApp(app, 200);
+      } catch (Throwable e) {
+        LOG.error("Got exception while deleting app", e);
+      }
+    }
+  }
+
+  private RunId runAndWait(Id.Program program) throws Exception {
+    LOG.info("Starting program {}", program);
+    startProgram(program);
+    waitState(program, ProgramRunStatus.RUNNING.toString());
+    return getRunId(program);
+  }
+
+  private void waitForStop(Id.Program program, boolean needsStop) throws Exception {
+    if (needsStop && getProgramStatus(program).equals(ProgramRunStatus.RUNNING.toString())) {
+      LOG.info("Stopping program {}", program);
+      stopProgram(program);
+    }
+    waitState(program, STOPPED);
+    LOG.info("Program {} has stopped", program);
+  }
+
   private RunId getRunId(Id.Program program) throws Exception {
-    List<RunRecord> runRecords = getProgramRuns(program, "RUNNING");
-    Assert.assertEquals(1, runRecords.size());
-    return RunIds.fromString(runRecords.iterator().next().getPid());
+    return getRunId(program, null);
+  }
+
+  private RunId getRunId(final Id.Program program, @Nullable final RunId exclude) throws Exception {
+    final AtomicReference<Iterable<RunRecord>> runRecords = new AtomicReference<>();
+    Tasks.waitFor(1, new Callable<Integer>() {
+      @Override
+      public Integer call() throws Exception {
+        runRecords.set(Iterables.filter(getProgramRuns(program, "RUNNING"),
+                                        new Predicate<RunRecord>() {
+                                          @Override
+                                          public boolean apply(RunRecord input) {
+                                            return exclude == null || !input.getPid().equals(exclude.getId());
+                                          }
+                                        }));
+        return Iterables.size(runRecords.get());
+      }
+    }, 60, TimeUnit.SECONDS, 10, TimeUnit.MILLISECONDS);
+    Assert.assertEquals(1, Iterables.size(runRecords.get()));
+    return RunIds.fromString(Iterables.getFirst(runRecords.get(), null).getPid());
+  }
+
+  private int status(org.apache.http.HttpResponse response) {
```

Comment:
Hmmm, I think I had used common-http's Http client in MetadataTestBase because at the time the org.apache one didn't allow to send request body in a delete call. Can you report a jira for me to standardize on Http client in these classes? Seems weird to have to  use two different ones.